Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,550 @@
+    let model = Model(date: date)
+    assertRoundTrip(model: model, encoded: ["date": date])
+  }
+
+  // Uncomment if we decide to reenable embedded DocumentReference's
+//
+//  func testDocumentReference() {
+//    struct Model: Codable, Equatable {
+//      let doc: DocumentReference
+//    }
+//    let d = FSTTestDocRef("abc/xyz")
+//    let model = Model(doc: d)
+//    assertRoundTrip(model: model, encoded: ["doc": d])
+//  }
+//
+//  // DocumentReference is not Codable unless embedded in a Firestore object.
+//  func testDocumentReferenceEncodes() {
+//    let doc = FSTTestDocRef("abc/xyz")
+//    do {
+//      _ = try JSONEncoder().encode(doc)
+//      XCTFail("Failed to throw")
+//    } catch FirebaseFirestoreSwift.FirestoreEncodingError.encodingIsNotSupported {
+//      return
+//    } catch {
+//      XCTFail("Unrecognized error: \(error)")
+//    }
+//  }
+
+  func testTimestamp() {
+    struct Model: Codable, Equatable {
+      let timestamp: Timestamp
+    }
+    let t = Timestamp(date: Date())
+    let model = Model(timestamp: t)
+    assertRoundTrip(model: model, encoded: ["timestamp": t])
+  }
+
+  func testBadValue() {
+    struct Model: Codable, Equatable {
+      let x: Int
+    }
+    let dict = ["x": "abc"] // Wrong type;
+    let model = Model(x: 42)
+    assertDecodingThrows(dict, encoded: model)
+  }
+
+  func testValueTooBig() {
+    struct Model: Codable, Equatable {
+      let x: CChar
+    }
+    let dict = ["x": 12345] // Overflow
+    let model = Model(x: 42)
+    assertDecodingThrows(dict, encoded: model)
+    assertRoundTrip(model: model, encoded: ["x": 42])
+  }
+
+  // Inspired by https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java
+  func testBeans() {
+    struct Model: Codable, Equatable {
+      let s: String
+      let d: Double
+      let f: Float
+      let l: CLongLong
+      let i: Int
+      let b: Bool
+      let sh: CShort
+      let byte: CChar
+      let uchar: CUnsignedChar
+      let ai: [Int]
+      let si: [String]
+      let caseSensitive: String
+      let casESensitive: String
+      let casESensitivE: String
+    }
+    let model = Model(
+      s: "abc",
+      d: 123,
+      f: -4,
+      l: 1_234_567_890_123,
+      i: -4444,
+      b: false,
+      sh: 123,
+      byte: 45,
+      uchar: 44,
+      ai: [1, 2, 3, 4],
+      si: ["abc", "def"],
+      caseSensitive: "aaa",
+      casESensitive: "bbb",
+      casESensitivE: "ccc"
+    )
+    let dict = [
+      "s": "abc",
+      "d": 123,
+      "f": -4,
+      "l": 1_234_567_890_123,
+      "i": -4444,
+      "b": false,
+      "sh": 123,
+      "byte": 45,
+      "uchar": 44,
+      "ai": [1, 2, 3, 4],
+      "si": ["abc", "def"],
+      "caseSensitive": "aaa",
+      "casESensitive": "bbb",
+      "casESensitivE": "ccc",
+    ] as [String: Any]
+
+    assertRoundTrip(model: model, encoded: dict)
+  }
+
+  func testCodingKeys() {
```

Comment:
A simpler test with two or three fields could verify that coding keys allows omitting fields.

Also, this test name should describe what about CodingKeys we're testing.

Reply:
Done.